### PR TITLE
Add tls connection for openstack and option "openstack-tls-ca-cert"

### PR DESCRIPTION
**Pull request by TrumanLing**
add tls connection for openstack and option "openstack-tls-ca-cert"

### CODE REVIEW

File: drivers/openstack/client.go [hunk 1]
```diff
@@ -439,12 +444,26 @@ func (c *GenericClient) Authenticate(d *Driver) error {
...
+		config = &tls.Config{InsecureSkipVerify: true}
+	} else {
+		flCa := filepath.Join(utils.GetMachineCertDir(), "ca.crt")
+		if d.TlsCACert != "" {
+			flCa = d.TlsCACert
+		}
+		certPool := x509.NewCertPool()
+		file, err := ioutil.ReadFile(flCa)
+		if err != nil {
+			log.Debug("Couldn't read ca cert %s: %s", flCa, err)
```
Comment (ggiamarchi) [inline drivers/openstack/client.go]: Here we should return the error instead of just log it.

File: drivers/openstack/client.go [hunk 2]
```diff
@@ -439,12 +444,26 @@ func (c *GenericClient) Authenticate(d *Driver) error {
...
 
+	// Configure custom TLS settings.
 	if d.Insecure {
-		// Configure custom TLS settings.
-		config := &tls.Config{InsecureSkipVerify: true}
-		transport := &http.Transport{TLSClientConfig: config}
-		provider.HTTPClient.Transport = transport
+		config = &tls.Config{InsecureSkipVerify: true}
+	} else {
+		flCa := filepath.Join(utils.GetMachineCertDir(), "ca.crt")
```
Comment (ggiamarchi) [inline drivers/openstack/client.go]: Why the default value `ca.crt` ? With your code the use of the certificate becomes mandatory. It's a mistake. In my opinion the code should look something like

```ruby
if d.Insecure {
	config = &tls.Config{InsecureSkipVerify: true}
}

if d.TlsCACert != "" {
	flCa := filepath.Join(utils.GetMachineCertDir(), d.TlsCACert)
	certPool := x509.NewCertPool()
	file, err := ioutil.ReadFile(flCa)
	if err != nil {
		return err
	}
	certPool.AppendCertsFromPEM(file)
	config = &tls.Config{
		RootCAs: certPool,
		InsecureSkipVerify: false,
	}
}

transport := &http.Transport{TLSClientConfig: config}
provider.HTTPClient.Transport = transport
```

**Comment by TrumanLing** [comment]
@ggiamarchi I am done. Thx for your review.

**Comment by ehazlett** [comment]
@ggiamarchi this look good to you?

**Comment by ggiamarchi** [comment]
@ehazlett Code looks good and tested ok.

@TrumanLing Can you please describe the new flag in [the documentation](https://github.com/docker/machine/blob/master/docs/drivers/openstack.md) ? Sorry, i forgot this point in my first review.

**Comment by ehazlett** [comment]
@ggiamarchi thanks!

**Comment by TrumanLing** [comment]
@ggiamarchi @ehazlett Could I provide another PR for describing the flag?

**Comment by ggiamarchi** [comment]
@TrumanLing I don't have any objection. I leave the final word to @ehazlett.

**Comment by dmp42** [comment]
@TrumanLing can you rather add documentation in this same PR, and amend/squash commits?

@ggiamarchi are we good otherwise?

Thanks for this!

**Comment by ggiamarchi** [comment]
@dmp42 Yes, appart the missing documentation it looks good to me.

**Comment by dmp42** [comment]
@TrumanLing you need a quick rebase - and doc - then that should be good to go.
Thanks a lot for this!

**Comment by dgageot** [comment]
Hi @TrumanLing any chance you could rebase this PR. Otherwise we might have to close it

**Comment by dgageot** [comment]
@TrumnLing It sounds like this PR is stalled. I'm going to close it. Please feel free to reopen it/rebase it if you'd like to carry on working on this PR. We'd be happy to review it!

**Pull request closed without merging**